File: app/src/main/java/com/imceits/aungtuntun/alephcodeassignment/data/ToolRepository.java
package com.imceits.aungtuntun.alephcodeassignment.data;

import androidx.lifecycle.LiveData;

import com.imceits.aungtuntun.alephcodeassignment.data.entity.Friends;
import com.imceits.aungtuntun.alephcodeassignment.data.entity.ToolsRent;
import com.imceits.aungtuntun.alephcodeassignment.data.pojos.ToolsDto;
import com.imceits.aungtuntun.alephcodeassignment.data.pojos.ToolsRentDto;

import java.util.List;
import java.util.concurrent.ExecutorService;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class ToolRepository {
    private final ToolsRentDao toolsRentDao;
    private final ExecutorService executorService;
    private LiveData<List<ToolsDto>> toolsLiveData;
    private LiveData<List<Friends>> friendsLiveData;

    @Inject
    public ToolRepository(ToolsRentDao toolsRentDao, ExecutorService executorService) {
        this.toolsRentDao = toolsRentDao;
        this.executorService = executorService;
        toolsLiveData = toolsRentDao.getAllTools();
        friendsLiveData = toolsRentDao.getAllFriends();
    }


    public void insertToolsRent(ToolsRent toolsRent) {
        executorService.execute(() -> {
            ToolsRent data = toolsRentDao.getExistDataById(toolsRent.getFriendId(), toolsRent.getToolId(), Status.BORROWED);
            if (data != null && data.getId() > 0) {
                toolsRentDao.updateExistBorrowedRecord(data.getId(), toolsRent.getToolId());
            }else {
                toolsRentDao.saveToolsRent(toolsRent);
            }

        });
    }

    public void updateToolsRent(ToolsRent toolsRent) {
        executorService.execute(() -> toolsRentDao.updateReturnedTools(toolsRent.getStatus(), toolsRent.getId()));
    }

    public LiveData<List<ToolsDto>> getAllTools() {
       return toolsLiveData;
    }

    public LiveData<List<Friends>> getAllFriends() {
        return friendsLiveData;
    }

    public LiveData<Friends> getFriendsData(int id) {
        return toolsRentDao.getFriendById(id);
    }

    public LiveData<List<ToolsRentDto>> getToolsRentList(int id) {
        return toolsRentDao.getRentToolsByFriendId(id, Status.BORROWED);
    }


}

File: app/src/main/java/com/imceits/aungtuntun/toolsbox/data/ToolRepository.kt
package com.imceits.aungtuntun.toolsbox.data

import androidx.lifecycle.LiveData
import com.imceits.aungtuntun.toolsbox.data.entity.Friends
import com.imceits.aungtuntun.toolsbox.data.entity.ToolsRent
import com.imceits.aungtuntun.toolsbox.data.views.ToolsDto
import com.imceits.aungtuntun.toolsbox.data.views.ToolsRentDto

class ToolRepository(private val toolsRentDao: ToolsRentDao) {

    val toolsList: LiveData<List<ToolsDto>> = toolsRentDao.getAllTools()
    val friendList = toolsRentDao.getAllFriends()

    suspend fun insertToolsRent(toolsRent: ToolsRent) {
        val data = toolsRentDao.getExistDataById(toolsRent.friendId, toolsRent.toolId, Status.BORROWED)
        data?.let {
            toolsRentDao.updateExistBorrowedRecord(data.id, toolsRent.toolId)
        }?: toolsRentDao.saveToolsRent(toolsRent)
        /*if (data?.id!! > 0){
            toolsRentDao.updateExistBorrowedRecord(data.id, toolsRent.toolId)
        }else{
            toolsRentDao.saveToolsRent(toolsRent)
        }*/
    }

    suspend fun updateToolsRent(toolsRent: ToolsRent) {
        toolsRentDao.updateReturnedTools(toolsRent.status, toolsRent.id)
    }

    fun getFriendsData(id: Int) : LiveData<Friends> {
        return toolsRentDao.getFriendById(id)
    }

    fun getToolsRentList(id: Int): LiveData<List<ToolsRentDto>>? {
        return toolsRentDao.getRentToolsByFriendId(id, Status.BORROWED)
    }

}